python/dgl/data/dgl_dataset.py

```diff
@@ -0,0 +1,157 @@
+"""Basic DGL Dataset
+"""
+
+from __future__ import absolute_import
+
+import os, sys
+from .utils import download, extract_archive, get_download_dir, makedirs
+from ..utils import retry_method_with_fix
+
+class DGLDataset(object):
+    r"""The Basic DGL Dataset for creating graph datasets.
+    This class defines a basic template class for DGL Dataset.
+    TODO(xiangsx):
+        Suport to_pytorch_dataset, to_mxnet_dataset, to_tensorflow_dataset
+        so DGL Dataset can easily converted to Framework specific Dataset
+    Parameters
+    name : str
+        Name of the dataset
+    url : str
+        Url to download the raw dataset
+    raw_dir : str
+        Raw file directory to download/contains the input data directory.
+        Default: ~/.dgl/
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    """
+    def __init__(self, name, url=None, raw_dir=None, force_reload=False):
+        self._name = name
+        self._url = url
+        self._force_reload = force_reload
+
+        # if no dir is provided, the default dgl download dir is used.
+        if raw_dir is None:
+            self._raw_dir = get_download_dir()
+
+        self._load()
+
+    def download(self):
+        r"""Downloads the dataset to the :obj:`self.raw_dir` folder.
+            Can be ignored if the dataset is already in self.raw_dir
+        """
+        pass
+
+    def save(selfs):
+        r"""Save the processed dataset into files.
+            Use dgl.utils.data.save_graphs to save dgl graph into files.
+            Use dgl.utils.data.save_info to save extra dict information into files.
```

Comment:
What are the typical `extra dict info` here? Graph classification may have graph-level labels that can't be included in `ndata`/`edata`.

Reply:
labels, for heterograph, there maybe target type, etc.

Reply:
Then are we gonna save/load the graph labels with another function?